cpp/common/src/codingstandards/cpp/rules/uncheckedrangedomainpoleerrors/UncheckedRangeDomainPoleErrors.qll

```diff
@@ -53,18 +54,95 @@ predicate hasDomainError(FunctionCall fc, string description) {
     functionWithDomainError = getMathVariants("log1p") and
     upperBound(fc.getArgument(0)) < -1.0 and
     description = "argument is less than 1"
+    or
+    functionWithDomainError = getMathVariants("fmod") and
+    fc.getArgument(1).getValue().toFloat() = 0 and
+    description = "y is 0"
   )
 }
 
+predicate hasRangeError(FunctionCall fc, string description) {
+  exists(Function functionWithRangeError | fc.getTarget() = functionWithRangeError |
+    functionWithRangeError.hasGlobalOrStdName(["abs", "labs", "llabs", "imaxabs"]) and
+    fc.getArgument(0) = any(MINMacro m).getAnInvocation().getExpr() and
+    description = "argument is most negative number"
+  )
+}
+
+predicate hasPoleError(FunctionCall fc, string description) {
+  exists(Function functionWithPoleError | fc.getTarget() = functionWithPoleError |
+    functionWithPoleError = getMathVariants("atanh") and
+    (
+      fc.getArgument(0).getValue().toFloat() = -1.0
```

Comment:
```suggestion
      (fc.getArgument(0).getValue().toFloat() + 1.0).abs() < 0.001
```

Due to the imprecision of floating points we can't compare them like this.
You need to subtract them and check whether their absolute value is smaller than some epsilon.

Reply:
discussed internally - conclusion:
quote Remco - "Since we are checking literals, similar to [https://github.com/github/codeql/blob/8b919148262f8dfd96d292f682d4f309692ae75b/jav[…]ode/java/dataflow/internal/rangeanalysis/SignAnalysisCommon.qll](https://github.com/github/codeql/blob/8b919148262f8dfd96d292f682d4f309692ae75b/java/ql/lib/semmle/code/java/dataflow/internal/rangeanalysis/SignAnalysisCommon.qll#L320) for which the floating point don’t have a gap, I guess the direct comparison is fine."